core/jobs/batch_jobs/voiceover_migration_job.py

```diff
@@ -0,0 +1,314 @@
+                    )
+
+                entity_voiceovers_object.add_new_content_for_voiceover(
+                    content_id)
+                entity_voiceovers_object.add_voiceover(
+                    content_id,
+                    feconf.VoiceoverType.MANUAL,
+                    manual_voiceover_instance
+                )
+
+                entity_voiceovers_id_to_entity_voiceovers[
+                    entity_voiceovers_id] = entity_voiceovers_object
+
+        return list(entity_voiceovers_id_to_entity_voiceovers.values())
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        """Returns a PCollection of results for the exploration for which an
+        exploration voiceover migration has been done.
+
+        Returns:
+            PCollection. A PCollection of results for the exploration for which
+            an exploration voiceover migration has been done.
+        """
+        exploration_models = (
+            self.pipeline
+            | 'Get exploration models' >> ndb_io.GetModels(
+                exp_models.ExplorationModel.get_all())
+        )
+
+        exploration_voice_artists_link_models = (
+            self.pipeline
+            | 'Get exploration voice artists link models' >> (
+                ndb_io.GetModels(
+                    voiceover_models.ExplorationVoiceArtistsLinkModel.get_all()
+                )
+            )
+        )
+
+        voice_artist_metadata_models = (
+            self.pipeline
+            | 'Get voice artist metadata model' >> (
+                ndb_io.GetModels(
+                    voiceover_models.VoiceArtistMetadataModel.get_all()
+                )
+            )
+        )
+
+        # Pair each exploration model with its exploration ID and create
+        # key-value pairs.
+        paired_exploration_models = (
+            exploration_models
+            | 'Pair Exploration ID to model' >> beam.Map(
+                lambda model: (model.id, model))
+        )
+
+        # Pair each exploration voice artists link model with its exploration
+        # ID and create key-value pairs.
+        paired_exploration_voice_artists_link_models = (
+            exploration_voice_artists_link_models
+            | 'Pair Exploration ID to voice artist link model' >> beam.Map(
+                lambda model: (model.id, model))
+        )
+
+        # Group the key-value pairs from both PCollections by the
+        # Exploration ID.
+        grouped_models = {
+            'exploration_model': paired_exploration_models,
+            'exploration_voice_artists_link_model': (
+                paired_exploration_voice_artists_link_models),
+        } | 'Group by Exploration ID' >> beam.CoGroupByKey()
+
+        entity_voiceovers_instances = (
+            grouped_models
+            | 'Filter invalid exploration and voice artist link models' >> (
+                beam.Filter(
+                    lambda element: (
+                        element[0] != '' and
```

Comment:
Why would the empty-string case occur?

Reply:
That's worrisome, we should fix that. Is there an issue tracking this? If not I think I need to see a link to an issue that outlines repro steps and a TODO here to drop this condition once the issue is fixed (if the code here lives beyond that).